### PR TITLE
Matter Water Sensor Device Integration

### PR DESCRIPTION
**Pull request by hcarter-775**
Check all that apply

# Type of Change

- [ ] WWST Certification Request
     - If this is your first time contributing code:
          - [x] I have reviewed the README.md file
          - [x] I have reviewed the CODE_OF_CONDUCT.md file
          - [x] I have signed the CLA
     - [ ] I plan on entering a WWST Certification Request or have entered a request through the WWST Certification console at developer.smartthings.com
- [ ] Bug fix
- [X] New feature
- [ ] Refactor

# Checklist

- [ ] I have performed a self-review of my code
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have verified my changes by testing with a device or have communicated a plan for testing
- [ ] I am adding new behavior, such as adding a sub-driver, and have added and run new unit tests to cover the new behavior

# Description of Change
This PR includes protoype support for:
* Water Leak Detector
* Water Freeze Detector
* Rain Sensor

# Summary of Completed Tests
N/A

### CODE REVIEW

**Review by ctowns — APPROVED** [review]
We'll need some profiles for this device type too!

File: drivers/SmartThings/matter-sensor/capabilities/customRainSensor.yml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+name: Custom Rain Sensor
+id: smilevirtual57983.customRainSensor
```
Comment (HunsupJung) [inline drivers/SmartThings/matter-sensor/capabilities/customRainSensor.yml:2]: We can use the [rainSensor](https://github.ecodesamsung.com/iot/SmartThingsCapabilities/blob/main/capabilities/rain-sensor.src/rain-sensor.yml) capability. right?

File: drivers/SmartThings/matter-sensor/src/init.lua [hunk 2]
```diff
@@ -333,11 +424,21 @@ local matter_driver_template = {
       clusters.SmokeCoAlarm.attributes.TestInProgress,
     },
     [capabilities.hardwareFault.ID] = {
-      clusters.SmokeCoAlarm.attributes.HardwareFaultAlert
+      -- clusters.SmokeCoAlarm.attributes.HardwareFaultAlert,
```
Comment (HunsupJung) [inline drivers/SmartThings/matter-sensor/src/init.lua:427]: I don't think it needs to exclude it.

File: drivers/SmartThings/matter-sensor/src/init.lua [hunk 3]
```diff
@@ -39,15 +39,53 @@ if version.api < 10 then
...
+          local this_id = dt.device_type_id
+          for name, info in pairs(BOOLEAN_DEVICE_TYPE_INFO) do
+              if this_id == info.id then
+                  device:set_field(name, ep.endpoint_id)
+              end
+          end
+      end
+  end
+  device:set_field(BOOLEAN_DEVICE_TYPES_CHECKED, 1, {persist = true})
+end
```
Comment (HunsupJung) [inline drivers/SmartThings/matter-sensor/src/init.lua:87]: I think you wanted to call this function only once after onboarding.
Can't driver call it only once at device_added function?

File: drivers/SmartThings/matter-sensor/src/init.lua [hunk 4]
```diff
@@ -221,6 +309,9 @@ local matter_driver_template = {
       [clusters.PressureMeasurement.ID] = {
         [clusters.PressureMeasurement.attributes.MeasuredValue.ID] = pressure_attr_handler,
       },
+      [clusters.BooleanStateConfiguration.ID] = {
+        [clusters.BooleanStateConfiguration.types.SensorFaultBitmap.ID] = sensor_fault_handler,
```
Comment (HunsupJung) [inline drivers/SmartThings/matter-sensor/src/init.lua:313]: clusters.BooleanStateConfiguration.attributes.SensorFault.ID

File: drivers/SmartThings/matter-sensor/src/init.lua [hunk 5]
```diff
@@ -333,11 +424,21 @@ local matter_driver_template = {
       clusters.SmokeCoAlarm.attributes.TestInProgress,
     },
     [capabilities.hardwareFault.ID] = {
-      clusters.SmokeCoAlarm.attributes.HardwareFaultAlert
+      -- clusters.SmokeCoAlarm.attributes.HardwareFaultAlert,
+      clusters.BooleanStateConfiguration.types.SensorFaultBitmap,
```
Comment (HunsupJung) [inline drivers/SmartThings/matter-sensor/src/init.lua:428]: clusters.BooleanStateConfiguration.attributes.SensorFault

Comment (hcarter-775) [reply drivers/SmartThings/matter-sensor/src/init.lua:428]: thanks!

File: drivers/SmartThings/matter-sensor/src/init.lua (hunk 4 above)
Comment (hcarter-775) [reply drivers/SmartThings/matter-sensor/src/init.lua:313]: thanks!

File: drivers/SmartThings/matter-sensor/src/init.lua (hunk 3 above)
Comment (hcarter-775) [reply drivers/SmartThings/matter-sensor/src/init.lua:87]: Yep, it's only being called in device_init

Comment (hcarter-775) [reply drivers/SmartThings/matter-sensor/src/init.lua:87]: Specifically if it hasn't been done before, per the `BOOLEAN_DEVICE_TYPES_CHECKED` gate

File: drivers/SmartThings/matter-sensor/src/init.lua (hunk 2 above)
Comment (hcarter-775) [reply drivers/SmartThings/matter-sensor/src/init.lua:427]: you are right, thank you!

File: drivers/SmartThings/matter-sensor/capabilities/customRainSensor.yml (hunk 1 above)
Comment (hcarter-775) [reply drivers/SmartThings/matter-sensor/capabilities/customRainSensor.yml:2]: at this point, I think we can? Not 100% sure.

Comment (HunsupJung) [reply drivers/SmartThings/matter-sensor/capabilities/customRainSensor.yml:2]: [rainSensor](https://github.ecodesamsung.com/iot/SmartThingsCapabilities/blob/main/capabilities/rain-sensor.src/rain-sensor.yml) was merged to production branch yesterday.

**Pull request merged**